IBM/ibm-common-service-operator
# Automatically clean up NSS CRs and operator in Simple TP or All ns mode

ticket: https://github.ibm.com/ibmprivatecloud/roadmap/issues/61040

### Context 
After upgrading from CS v3 to CS v4, the v3 Namespace Scope operator and CRs will be expected to be cleaned up by the common service operator automatically in Simple Topology or AllNamespace mode as they are no longer needed.

### NameSpace Scope clean up Workflow
1. CS operator upgrades to v4.
2. Check if all Bedrock v3 core services are no longer requested by CloudPak by inspecting the v4 OperandRegistry.
3. Check the deployment topology, whether it is a Simple Topology or in All Namespace Mode.
4. Delete v3 NSS operator in the servicesNamespace firstly if they still remain to prevent adding the ownerReferences back to `namespace-scope` configMap
5. Remove the ownerReferences in `namespace-scope` configMap in case it is deleted with `common-service` NamespaceScope CR and crashes cs operator pod
6. Lastly, delete all the remaining NamespaceScope CRs

### CS reconciliation
1. The CS operator will check if there are still `no-op` operators requested in the OperandRegistry. If there are none, it will trigger the reconciliation to clean up all NSS resources. If there are still some operators, it will skip the reconciliation.

### How to test
Test image: quay.io/yuchen_shen/cs_operator:nss_cleanup

- Simple TP
   1. Install a v3 CPFS in a single-shared cluster in one namespace `ibm-common-services`, or in a dedicated cluster in a custom namespace.
   3. Upgrade from v3 to v4 and configure the operator and service namespace in the CS CR.
   4. Apply the test image in the v4 CS CSV.
   5. Check that all NSS CRs and NSS operators in the service namespace are cleaned up automatically.

- All Namespaces
   1. Install a v3 CPFS and perform the upgrade.
   2. Apply the test image.
   3. Check if NSS relations are all gone.

- Advanced TP
   1. Install a v3 CPFS and upgrade to Advanced TP.
   2. Apply the test image.
   3. Ensure that NSS relations remain in the cluster.
